feat(gatsby-remark-prismjs): add in ability to use comments to specify line highlighting

phacks opened the pull request:
Closes #9235 

Implements the following comments directive to declare lines to highlights in `gatsby-remark-prismjs`:

- `highlight-line` highlights the current line;
- `highlight-next-line` highlights the next line;
- `highlight-start` highlights the lines until the matching `hightlight-end`;
- `highlight-range{1, 4-6}` will highlight the next line, and the fourth, fifth and sixth lines.

Also refactors `gatsby-remark-embed-snippet` to rely on this implementation of the directive instead of a local one.

Thanks @DSchau for the headstart :)

DSchau commented on packages/gatsby-remark-prismjs/README.md:
```diff
@@ -228,6 +228,29 @@ You can also add line highlighting. It adds a span around lines of code with a
+      resolve: `gatsby-transformer-remark`, //highlight-line
+      options: {
+        // highlight-start
+        plugins: [`gatsby-remark-prismjs`],
```
maybe a different example? I could see people getting confused that this is somehow related to actual configuration they have to do!

DSchau commented on packages/gatsby-remark-prismjs/README.md:
```diff
@@ -228,6 +228,29 @@ You can also add line highlighting. It adds a span around lines of code with a
+- `highlight-line` highlights the current line;
+- `highlight-next-line` highlights the next line;
+- `highlight-start` highlights the lines until the matching `hightlight-end`;
+- `highlight-range{1, 4-6}` will highlight the next line, and the fourth, fifth and sixth lines.
```
I'm not actually sure this is an improvement! I'd almost rather we cut this out. What do you think?

phacks replied:
Added it for feature parity with `gatsby-remark-embed-snippet` (see [this bit of docs](https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby-remark-embed-snippet#yaml-example)), however I reckon it can be quite confusing. Should we remove it from the docs and keep it in the code, so as not to break the embed-snippet plugin?

DSchau replied:
ah! Yeah, I suppose if we're providing the feature we may as well document it!

Thanks for catching that!

phacks replied:
Adding it back in! Will look for a proper example though, so as to make all directives as clear as possible :)

DSchau commented on packages/gatsby-remark-prismjs/highlight-line-range.js:
```diff
@@ -0,0 +1,160 @@
+"use strict"
```
This file shouldn't be committed! We'll want to remove

phacks replied:
oopsie, removed 😅

phacks commented:
@DSchau Thanks so much! Updated the PR to reflect your early comments, and your visual testing suite made me realize I had a bug in the old range behavior (now fixed 😇).

I added two more tests to the kitchen sink, and will have another look into your comments in the next few days!

DSchau commented:
@phacks I think that bug was present in my version, so... thanks for fixing :) Glad to hear it was helpful!

phacks commented:
@DSchau FYI I tried running a local version of reactjs.org (https://github.com/reactjs/reactjs.org) with the changes in this PR, and all seems to run smoothly 😊

![image](https://user-images.githubusercontent.com/2587348/48126072-8e8b8680-e280-11e8-8990-469040e83f15.png)

phacks commented:
@DSchau Hi! Do you think you might have some time to have another look at this PR in the coming days? Thanks!

DSchau commented:
@phacks definitely! I'll try and get to it this afternoon!

DSchau commented:
Testing this out now 🤞 Going to check with reactjs.org, the little kitchen sink starter, and then that should be enough evidence it hasn't caused a regression!

DSchau commented on packages/gatsby-remark-prismjs/src/highlight-line-range.js:
```diff
@@ -0,0 +1,147 @@
+const END_DIRECTIVE = /highlight-end/
+const plainTextWithLFTest = /<span class="token plain-text">[^<]*\n[^<]*<\/span>/g
+
+const stripComment = line =>
```
I think we'll have to re-visit this. Here's what happens with a JSX comment:

![screen shot 2018-11-12 at 2 19 51 pm](https://user-images.githubusercontent.com/3924690/48372997-0ba26b80-e686-11e8-9c7d-1c0c6e6d9506.png)

So it's _close_ but not quite there! Checking out the reactjs.org site now.

phacks replied:
Fixed in 4b80fc7 — problem was that the code that was processed in the `stripComment` function had already gone through Prism, so there was a few extra `span`s here and there.

I added this comment for clarity:

https://github.com/gatsbyjs/gatsby/blob/4b80fc7291e659e79f8f9aef209865bdd5e69656/packages/gatsby-remark-prismjs/src/highlight-line-range.js#L3-L19

Not 100% happy with this solution though, if you have an idea for a more elegant one I’d be willing to try it!

DSchau commented:
@phacks sounds great! Please do let one us know if we can help!

phacks commented:
> Found another (small?) bug. Looks like it's dropping lines of code unnecessarily sometimes, consider the following snippet
> 
> ```
> ```jsx
> class Square extends React.Component {
>   render() {
>     // highlight-range{2-4}
>     return (
>       <button className="square">
>         {this.props.value}
>       </button>
>     );
>   }
> }
> ```
> ```
> it's rendered _mostly_ correctly, but seems to drop the tailing parenthesis.
> 
> ![screen shot 2018-11-12 at 2 27 47 pm](https://user-images.githubusercontent.com/3924690/48373338-25907e00-e687-11e8-9487-418cfa0701c3.png)

Should be fixed in 003cd54!

phacks commented:
@DSchau Rebased the PR from master, fixed both bugs and added a few comments — should be ready for you to review when you have some time 😄

DSchau commented:
@phacks will take another deep dive check on this today! Thanks for your patience!

DSchau commented on packages/gatsby-remark-prismjs/README.md:
````diff
@@ -228,6 +228,46 @@ You can also add line highlighting. It adds a span around lines of code with a
+
+  ```javascript
+  function calculateWinner(squares) {
+    //highlight-line
````
What would this actually do? Probably want to clean up this example a tad.

phacks replied:
Gave it another spin in 3ebbdbd, should make it more clear/explicit

DSchau commented on packages/gatsby-remark-prismjs/src/highlight-line-range.js:
````diff
@@ -0,0 +1,165 @@
+
+  // If a highlight range is passed with the language declaration, e.g.
+  //     ```jsx{1, 3-4}
+  // we only use that and do not try to parse highlight directives
````
Good work/idea to comment this! Important to keep this behavior for posterity.

phacks commented:
@DSchau Thanks! If I didn’t miss anything, by the next weekend I’ll try to : 

- Make sense of the snapshot updates in `gatsby-remark-embed-snippet`;
- Update the example in the README to be more explicit

Do you see anything else?

And by the way, I was talking to @tychota yesterday, who told me he looked forward for this being merged for a reason I hadn’t think of yet: IDE highlighting of code blocks.

Indeed, in a Markdown file, 

````
```css{1-3}
body {
  font-size: 14px;
}
```
````

is not recognized as CSS by (at least) VSCode, so there are no syntax coloration in this code block. Highlighting directives solves that problem too! 😊

phacks commented on packages/gatsby-remark-embed-snippet/src/__tests__/__snapshots__/index.js.snap:
```diff
@@ -260,7 +260,7 @@ Object {
         <pre class=\\"language-html\\"><code><span class=\\"token tag\\"><span class=\\"token tag\\"><span class=\\"token punctuation\\">&lt;</span>html</span><span class=\\"token punctuation\\">></span></span>
 <span class=\\"token tag\\"><span class=\\"token tag\\"><span class=\\"token punctuation\\">&lt;</span>body</span><span class=\\"token punctuation\\">></span></span>
-<span class=\\"gatsby-highlight-code-line\\"><span class=\\"token tag\\"><span class=\\"token tag\\"><span class=\\"token punctuation\\">&lt;</span>h1</span><span class=\\"token punctuation\\">></span></span>highlighted<span class=\\"token tag\\"><span class=\\"token tag\\"><span class=\\"token punctuation\\">&lt;/</span>h1</span><span class=\\"token punctuation\\">></span></span>
+<span class=\\"gatsby-highlight-code-line\\"><span class=\\"token tag\\"><span class=\\"token tag\\"><span class=\\"token punctuation\\">&lt;</span>h1</span><span class=\\"token punctuation\\">></span></span>highlighted<span class=\\"token tag\\"><span class=\\"token tag\\"><span class=\\"token punctuation\\">&lt;/</span>h1</span><span class=\\"token punctuation\\">></span></span> <span class=\\" ...
```
Indeed! And it led me to fix a bug where `<!-- highlight-line -->` wasn’t properly stripped down (see 10aee65). I updated the kitchen sink repo to include this example.

phacks commented:
@DSchau Hi! Here’s what I did:

- updated the README example
- fixed a bug (see [this comment](https://github.com/gatsbyjs/gatsby/pull/9696#discussion_r236038979))
- dug a little deeper into the remaining `gatsby-remark-embed-snippet` snapshot updates.

There are two updates remaining, one benign, the other one a small regression that will be fixed in an upcoming version of PrismJS:

### Snapshot update 1

```diff
--- a/packages/gatsby-remark-embed-snippet/src/__tests__/__snapshots__/index.js.snap
+++ b/packages/gatsby-remark-embed-snippet/src/__tests__/__snapshots__/index.js.snap
@@ -443,7 +443,7 @@ Object {
           "type": "html",
           "value": "<div class=\\"gatsby-highlight\\">
         <pre class=\\"language-jsx\\"><code><span class=\\"token tag\\"><span class=\\"token tag\\"><span class=\\"token punctuation\\">&lt;</
- <span class=\\"gatsby-highlight-code-line\\"><span class=\\"token plain-text\\"></span><span class=\\"token tag\\"><span class=\\"token tag\\"><span class=\\"token punctuation\\">&lt;</span>button</span><span class=\\"token punctuation\\">></span></span><span class=\\"token plain-text\\">Add Item</span><span class=\\"token tag\\"><span class=\\"token tag\\"><span class=\\"token punctuation\\">&lt;/</span>button</span><span class=\\"token punctuation\\">></span></span><span class=\\"token plain-text\\"></span>
+ <span class=\\"gatsby-highlight-code-line\\"><span class=\\"token plain-text\\"></span><span class=\\"token tag\\"><span class=\\"token tag\\"><span class=\\"token punctuation\\">&lt;</span>button</span><span class=\\"token punctuation\\">></span></span><span class=\\"token plain-text\\">Add Item</span><span class=\\"token tag\\"><span class=\\"token tag\\"><span class=\\"token punctuation\\">&lt;/</span>button</span><span class=\\"token punctuation\\">></span></span><span class=\\"token plain-text\\"> </span><span class=\\"token plain-text\\"></span>
```

**Cause**: Couldn’t pinpoint it clearly, seems a combination of JSX’s messy comment syntax once it went through PrismJS, and the hacky handling of plain-text spans with line separators
**Impact**: Seems negligible, it only adds an extra empty span when a `{/* highlight-line */}` is directly followed by a new line. Invisible to the end-user.

### Snapshot update 2

```diff
--- a/packages/gatsby-remark-embed-snippet/src/__tests__/__snapshots__/index.js.snap
+++ b/packages/gatsby-remark-embed-snippet/src/__tests__/__snapshots__/index.js.snap
@@ -875,7 +875,7 @@ Object {
           },
           "type": "html",
           "value": "<div class=\\"gatsby-highlight\\">
- <pre class=\\"language-yaml\\"><code><span class=\\"gatsby-highlight-code-line\\"><span class=\\"token key atrule\\">foo</span><span class=\\"token punctuation\\">:</span> <span class=\\"token string\\">\\"highlighted\\"</span>
+ <pre class=\\"language-yaml\\"><code><span class=\\"gatsby-highlight-code-line\\"><span class=\\"token key atrule\\">foo</span><span class=\\"token punctuation\\">:</span> \\"highlighted\\"
```

**Cause**: In `prismjs@1.15.0`, YAML strings followed by comments are not properly highlighted. See issue https://github.com/PrismJS/prism/issues/1601 and PR https://github.com/PrismJS/prism/pull/1602. It’s been merged, however it’s pending the release of `prismjs@1.16.0`, which hasn’t been announced yet.
**Impact**: Causes a small regression in `gatsby-remark-embed-snippet`. In the screenshot below, see how `highlighted` is not colorized properly:

![image](https://user-images.githubusercontent.com/2587348/48971158-7a80ac80-f016-11e8-9a7e-e1bd4f3ec268.png)

It can be mitigated by using a highlighting directive other than `highlight-line`.

---

As I believe this new handling of highlighting will be very beneficial to Gatsby developers, what I’d recommend is the following:

- If you’re OK with the overall PR, we do merge it;
- I file an issue in the Gatsby repo for the YAML colorization problem indicating it’s pending a PrismJS update and pointing out that another directive circumvents the problem;
- I file an issue in PrismJS to see if they may give some timeline over the release of `1.16.0`. 

Would you be OK with this?

Thanks again for your mighty patience reviewing this PR 😇

DSchau commented:
```
Successfully published:
 - gatsby-remark-embed-snippet@3.1.0
 - gatsby-remark-prismjs@3.1.0
```

tychota commented:
> And by the way, I was talking to @tychota yesterday, who told me he looked forward for this being merged for a reason I hadn’t think of yet: IDE highlighting of code blocks.

Oh yeah  !!

![image](https://user-images.githubusercontent.com/13785185/49022096-59de6100-f194-11e8-85d5-bcdd30112ab8.png)

🎉⭐️

baerrach commented:
> @phacks that's a nice win I hadn't thought of either! I see very little downsides of encouraging everyone to use `highlight-start` and `highlight-end`. I think it's a big improvement!

This reply probably belongs in its own issue, but I'll start here.

The documentation currently doesn't tell you how to disable highlighting.

DSchau's [kitchen sink](https://github.com/DSchau/gatsby-line-highlighting-kitchen-sink/blob/master/content/kitchen-sink.md) has the same types of example as [Gatsby Style Guide - Code formatting: Line highlighting](https://www.gatsbyjs.org/contributing/gatsby-style-guide/#code-formatting-line-highlighting) but the Gatsby site has both code blocks highlighted.

This may be related to related to https://github.com/gatsbyjs/gatsby/issues/11334#issuecomment-458064773 and using gatsby-remark-prismjs "^3.0.2". I've tried locally to upgrade to "^3.3.13" but the problem still persists.

The four backtick for code blocks should also be documented. I couldn't find that anywhere when looking at markdown formating of code blocks that included code blocks.

I can hack around the problem by using line range directives, for example:

    ````{0}
    ```javascript:title=gatsby-config.js
    module.exports = {
        siteMetadata: {
            title: `GatsbyJS`, // highlight-line
            siteUrl: `https://www.gatsbyjs.org`,
        },
    }
    ```
    ````

Does anyone know why the four backticks are not disabling the `highlight-line` directive?
